bitnami/etcd/3.5/debian-12/rootfs/opt/bitnami/scripts/libetcd.sh

```diff
@@ -572,6 +520,94 @@ recalculate_initial_cluster() {
     fi
 }
 
+########################
+# Remove the old member from the cluster if it exists
+# Globals:
+#   ETCD_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+remove_old_member_if_exist() {
+    local old_member_id
+    old_member_id=$(get_member_id)
+    if ! is_empty_value "$old_member_id"; then
+        info "Removing old member $old_member_id"
+        local -a extra_flags
+        read -r -a extra_flags <<<"$(etcdctl_auth_flags)"
+        is_boolean_yes "$ETCD_ON_K8S" && extra_flags+=("--endpoints=$(etcdctl_get_endpoints)")
+        etcdctl member remove "$old_member_id" "${extra_flags[@]}"
+    fi
+}
+
+########################
+# Add this member as a new member to the cluster
+# Globals:
+#   ETCD_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+add_new_member() {
+    local -a extra_flags
+    info "Adding new member to existing cluster"
+    read -r -a extra_flags <<<"$(etcdctl_auth_flags)"
+    is_boolean_yes "$ETCD_ON_K8S" && extra_flags+=("--endpoints=$(etcdctl_get_endpoints)")
+    extra_flags+=("--peer-urls=$ETCD_INITIAL_ADVERTISE_PEER_URLS")
+    mkdir -p "$(dirname $ETCD_NEW_MEMBERS_ENV_FILE)" || true
+    etcdctl member add "$ETCD_NAME" "${extra_flags[@]}" | grep "^ETCD_" >"$ETCD_NEW_MEMBERS_ENV_FILE"
+    replace_in_file "$ETCD_NEW_MEMBERS_ENV_FILE" "^" "export "
+    sync -d "$ETCD_NEW_MEMBERS_ENV_FILE"
+}
+
+########################
+# Check that this node is still a member of the cluster
+# Globals:
+#   ETCD_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+is_membership_intact() {
+    local tmp_file
+    local start_command=("etcd")
+    local pid
+    local ret=0
+
+    tmp_file=$(mktemp)
+
+    am_i_root && start_command=("run_as_user" "$ETCD_DAEMON_USER" "${start_command[@]}")
+    [[ -f "$ETCD_CONF_FILE" ]] && start_command+=("--config-file" "$ETCD_CONF_FILE")
+    $start_command > >(tee -a "$tmp_file") 2>&1 &
+    pid=$!
+    debug "Started etcd in background with PID $pid"
+    
+    while read -r line; do
+        echo "$line" # Stream the output
+        if [[ "$line" =~ (established TCP streaming connection with remote peer|the member has been permanently removed from the cluster|ignored streaming request; ID mismatch|\"error\":\"cluster ID mismatch\") ]]; then
+            kill "$pid"
+            wait "$pid" 2>/dev/null
+            debug "Stopped etcd"
+            break
+        fi
+    done < <(tail -f "$tmp_file")
+
+    if grep -q "the member has been permanently removed from the cluster\|ignored streaming request; ID mismatch" "$tmp_file"; then
+        info "The remote member ID is different from the local member ID"
+        ret=1
+    elif grep -q "\"error\":\"cluster ID mismatch\"" "$tmp_file"; then
+        info "The remote cluster ID is different from the local cluster ID"
+        ret=1
+    else
+        info "The member is still part of the cluster"
+    fi
+    rm -f "$tmp_file"
+    return $ret
```

Comment:
You can simplify this:

```suggestion
    trap rm -f "$tmp_file"
    if grep -q "the member has been permanently removed from the cluster\|ignored streaming request; ID mismatch" "$tmp_file"; then
        info "The remote member ID is different from the local member ID"
        return 1
    elif grep -q "\"error\":\"cluster ID mismatch\"" "$tmp_file"; then
        info "The remote cluster ID is different from the local cluster ID"
        return 1
    fi
    
    info "The member is still part of the cluster"
    return 0
```

Also, wouldn't it be better to rely on some etcdctl command to check if the node is still a member instead of starting etcd in background?

Reply:
If there is an etcdctl command that checks the data dir directly then I'm not aware of it. Most commands only work on running etcd endpoints.

Reply:
I have refactored as you suggested. Please check again.